torch/optim/asgd.py

```diff
@@ -34,7 +34,7 @@ def __init__(
         if not 0.0 <= weight_decay:
             raise ValueError(f"Invalid weight_decay value: {weight_decay}")
 
-        if foreach is False and capturable:
+        if foreach is False and capturable and not is_compiling():
```

Comment:
So single Tensor capturable is fine when we're compiling? :thinking:

Reply:
Basically this error is to prevent a user from setting capturable and running cudagraphs expecting it to be correct when it won't be. If we're compiling we detect the invalid behavior and don't run cudagraphs. (ie getitem and cpu tensors)

@janeyx99 it might be better to just check if capturable is true and cudagraphs is capturing for all optimizers. That way we could run all of the tests.

Reply:
Yea, but there's no real use case where someone specifies capturable in eager _without_ intending to run cuda graphs other than for testing, which doesn't count as a use case. Today, `capturable` in eager means `for cuda graphs` so the tests shouldn't mislead.